### PR TITLE
Zfs refactor

### PR DESCRIPTION
**Pull request by mbruzek**
Sorry for the excessive commits, it was an iterative process, but here is the gist:  
- I added tox.ini for tests and a Makefile to clean up after tests.
- Changed the ZfsPool class to always require a mount point.
- Re-factored the ZfsPool class to use the add method to add or create pools.
- Added back in the missing BtrfsPool class.
- Added comments in storage.py and fixed the logic to work with the ZfsPool class.

This has been tested successfully on juju 1.25 using Amazon. Running tox now lints and runs both tests which all pass.

### CODE REVIEW

File: lib/zfs.py [hunk 1]
```diff
@@ -1,51 +1,50 @@
...
+    and parity is striped across all disks within a raidz group.'''
 
     # The devices should be a list of strings.
     devices = []
     # The pool name must begin with a letter, and can only contain alphanumeric
     # characters as well as underscore ("_"), dash ("-"), period ("."), colon
     # (":"), and space (" "). The pool names "mirror", "raidz", "spare" and
     # "log" are reserved, as are names beginning with  the  pattern  "c[0-9]".
-    pool_name = 'zfs-storage-pool'
+    pool_name = 'juju-zfs-pool'
```
Comment (marcoceppi) [inline lib/zfs.py]: Should `pool_name` be configurable within `__init__`? ie

``` python
def __init__(self, mount_point, pool_name='juju-zfs-pool'):
```

File: lib/zfs.py [hunk 2]
```diff
@@ -55,35 +54,34 @@ def size(self):
...
+            cmd = 'sudo zpool add {0} {1} raidz '.format(force_flag,
+                                                         self.pool_name)
             cmd += ' '.join(devices)
             print(cmd)
             check_call(split(cmd))
-        else:
-            # The command that creates a zfs pool without a mount point.
-            cmd = 'sudo zpool create {0} raidz '.format(self.pool_name)
+            self.devices.append(devices)
+        else:  # The pool does not yet exist, so we must create a new pool.
```
Comment (marcoceppi) [inline lib/zfs.py:72]: It might be better to have a `_zpool_create` and `_zpool_add` private class methods to make unit testing easier and condense the complexity of this method.

File: lib/zfs.py [hunk 3]
```diff
@@ -55,35 +54,34 @@ def size(self):
...
+            # The command that creates a new raidz zfs disk pool.
+            cmd = 'sudo zpool create {0} -m {1} {2} raidz '.format(
```
Comment (marcoceppi) [inline lib/zfs.py:74]: You're always root, or should always assume root, for charms

File: lib/zfs.py (hunk 1 above)
Comment (mbruzek) [inline lib/zfs.py]: I removed the pool_name because there is no use case for naming pools differently, so I decided not to provide the option. I could add it back in if you think it is necessary, or can give me a use case where we would need that.

File: lib/zfs.py (hunk 3 above)
Comment (mbruzek) [inline lib/zfs.py:74]: My functional tests will not pass if I don't use sudo

**Comment by marcoceppi** [comment]
I have a few other questions, but I like the progress. I'll open issues for some of the other questions that don't pertain to this merge

File: lib/zfs.py (hunk 1 above)
Comment (marcoceppi) [inline lib/zfs.py]: I can see there a use case where based on the storage that was added a different zpool might want to be created. For instance, if I have two stoarge pools in metadata: `data` and `logs` These might be two unique pools with their own disk setups. The move to init would support these cases while still having the standard implementation of one pool present

File: lib/zfs.py (hunk 3 above)
Comment (marcoceppi) [inline lib/zfs.py:74]: bleh, that sucks. I'd say you could try to determine if root already, but it's not worth it for this pass.

File: layer.yaml [hunk 4]
```diff
@@ -9,3 +9,9 @@ defines:
       This is a build time option for the storage driver to handle the devices
       available to the charm. Only the 'overlay', 'aufs', 'btrfs', and 'zfs'
       are valid at this time.
+  mount-path:
```
Comment (lazypower) [inline layer.yaml]: i thought this was changing to mount-point?

File: lib/zfs.py (hunk 2 above)
Comment (lazypower) [inline lib/zfs.py:72]: +1 to that

File: lib/zfs.py (hunk 3 above)
Comment (lazypower) [inline lib/zfs.py:74]: I would argue that the test runner should be executed then as root so your test env is as close to your target env as possible, which negates the need for sudo inline here.

File: reactive/storage.py [hunk 5]
```diff
@@ -48,7 +58,7 @@ def install_storage_tools():
 @when('disk-pool-storage-attached', 'btrfs-tools-installed')
 def handle_btrfs_pool():
     '''The btrfs tools are installed, use the btfs tools.'''
-    mount_path = '/var/lib/docker'
```
Comment (lazypower) [inline reactive/storage.py:51]: big +1 to removing hard coded paths in the code

File: layer.yaml (hunk 4 above)
Comment (mbruzek) [inline layer.yaml]: Good catch, that is on an not yet pushed merge.

File: lib/zfs.py (hunk 1 above)
Comment (mbruzek) [inline lib/zfs.py]: ack, will add it

File: lib/zfs.py (hunk 3 above)
Comment (mbruzek) [inline lib/zfs.py:74]: Hey if we could do that I am all for it, I would like to make that a separate pull request.

File: Makefile [hunk 6]
```diff
@@ -0,0 +1,6 @@
+test:
+	tox
+clean:
```
Comment (lazypower) [inline Makefile:3]: minor suggestion to change this to a find stanza

find -name *.pyc -delete 

tomatoe/tomatoe

Comment (mbruzek) [inline Makefile:3]: I didn't know how to use the find command like this. I will do that next time. eh!

**Comment by lazypower** [comment]
+1 LGTM  as a submission, tests still need some love and docs, but overall this is shaping up nicely
